packages/utreexo/src/stump/proof.cairo

```diff
@@ -27,3 +30,170 @@ impl UtreexoBatchProofDisplay of Display<UtreexoBatchProof> {
         Result::Ok(())
     }
 }
+
+#[generate_trait]
+pub impl UtreexoBatchProofImpl of UtreexoBatchProofTrait {
+    /// Computes a set of roots given a proof and leaves hashes.
+    fn compute_roots(
+        self: @UtreexoBatchProof, del_hashes: Span<felt252>, num_leaves: u64,
+    ) -> Result<Span<felt252>, ByteArray> {
+        // Where all the parent hashes we've calculated in a given row will go to.
+        let mut calculated_root_hashes: Array<felt252> = array![];
+        // Target leaves
+        let mut leaf_nodes: Array<(u64, felt252)> = array![];
+
+        // Append targets with their hashes.
+        let mut i = 0;
+        while i != (*self.targets).len() {
```

Comment:
I think you can do zip with two pops, like this:

```rust
let mut positions = self.targets;
while let Option::Some(lhs) = del_hashes.pop_front() {
   if let Option::Some(rhs) = positions.pop_front() { ... }
}
```

It's also important to iterate over hashes rather than proof targets, because we cannot guarantee that prooof is correct.

Probably makes sense to create a generic `utils::zip<T, U>` implementation

Reply:
When implementing a generic function? Will it be used somewhere else in the codebase ?

Reply:
Not necessarily, but it's a useful primitive (since array doesn't provide anything similar out of the box).
Let's keep it for later